R/dataset.R

```diff
@@ -88,6 +88,80 @@ setMethod("notes<-", "CrunchDataset", function (x, value) {
     invisible(setEntitySlot(x, "notes", value))
 })
 
+#' Get and set the market size for Crunch datasets
+#'
+#' Crunch Datasets allow you to set a target population size in order to extrapolate
+#' population estimates from survey percentages. This function lets you work with
+#' the population size and magnitude.
+#'
+#' @param x a Crunch Dataset
+#' @param value For the setters, the `size` or `magnitude` to be set
+#' @param size the target population size, to remove a population set to `NULL`
+#' @param magnitude the order of magnitude with which to display the population
+#' size. Must be either `3`, `6`, or `9` for thousands, millions, and billions respectively.
+#' @return `popSize` and `popMagnitude` return the population size or
+#' magnitude. `setPopulation` returns the modified dataset.
+#' @name population
+#' @aliases popSize popMagnitude setPopulation popSize<- popMagnitude<-
+NULL
+
+#' @rdname population
+#' @export
+setMethod("popSize", "CrunchDataset", function (x) {
+    size <- settings(x)$population$size
+    if (size == 0) {
+        return(NULL)
+    }
+    return(size) #the API value 0 stands for "No Population set"
+})
+
+#' @rdname population
+#' @export
+
+setMethod("popSize<-", "CrunchDataset", function (x, value) {
+    setPopulation(x, size = value)
+})
+
+#' @rdname population
+#' @export
+setMethod("popMagnitude", "CrunchDataset", function (x) {
+    return(settings(x)$population$magnitude)
+})
+
+#' @rdname population
+#' @export
+setMethod("popMagnitude<-", "CrunchDataset", function (x, value) {
+   setPopulation(x, magnitude = value)
+})
+
+#' @rdname population
+#' @export
+setMethod("setPopulation", "CrunchDataset", function (x, size, magnitude) {
+    pop <- settings(x)$population
+    if (missing(magnitude)) {
+        if (is.null(pop$magnitude)) {
+            warning("Dataset magnitude not set, defaulting to thousands")
+            magnitude <- 3
+        } else {
+            magnitude <- pop$magnitude
+        }
+    }
+    if (missing(size)) {
+        if (is.null(pop$size)) {
+            halt("Dataset does not have a population, please set one before attempting to change magnitude")
+        }
+        size <- pop$size
+    }
+    if (is.null(size)) {
+        size <- 0
+    }
+    if (is.null(magnitude) || !(magnitude %in% c(3, 6, 9))){
+        halt("Magnitude must be either 3, 6, or 9")
+    }
+    settings(x)$population <- list(magnitude = magnitude, size = size)
+    invisible(return(x))
+})
```

Comment:
This looks pretty complicated (although it looks like accomplishes what we want).

One alternative to clean this up, that would also follow a couple of other patterns would be to have popSize and popMagnitude set their respective portions separately. This would make the `PATCH`es a bit cleaner because they would be only changing the one thing that changed rather than resending the same value.

Reply:
That's a good idea. The reason I went this route is that you have to send both size and magnitude together if there's no population set on the dataset. So that means that we need to send them both regardless, and I thought it was simpler to do it this way.